### DISCUSSION TITLE
How the TOTP secret key is stored in Keycloak ?

### DISCUSSION THREAD

**Discussion by farnulfo** [keycloak/keycloak#23207] [Q&A]
Hi,

The TOTP secret key for a user need to be accessible in plaintext to generate the TOTP number with the current time.
I would like to know how is it stored securely in keycloak ?

**Comment by farnulfo** [comment]
The TOTP secret key for a user is stored in table CREDENTIALS, SECRET_DATA column in json format:
Example: 
"SECRET_DATA" : "{\"value\":\"PjgA79HiQkU0ozMEWKV1\"}"
Column CREDENTIAL_DATA contains metadata about TOTP:
"CREDENTIAL_DATA" : "{\"subType\":\"totp\",\"digits\":6,\"counter\":0,\"period\":30,\"algorithm\":\"HmacSHA1\"}",

The secret key in SECRET_DATA column seems not be to encrypted with the user password because when the user change his password, SECRET_DATA value is not changed. May be to be able for a user to reset his password by receiving a link in a email: in this case the old password is not available to decryp the secret key.

The value in SECRET_DATA seems to be base64 encoded but the decodec value doesn't match the TOTP displayed in QR Code for the user in base32 encoding.

**Reply by SatriyoJati** [reply]
hi , where is it ? i didnt find any of this. My setup use postgresql as user database. I have created user and set the newly created user to use TOTP, I need to know where the secret key stored. My goal is to make apis that can access and share the credetntials, so other device can generate OTP.

**Reply by darius-m** [reply]
@SatriyoJati, you can simply add more OTP tokens if you want to add devices that can generate TOTP for an account. Extracting the information from the database directly would likely increase your attack surface by quite a bit and Keycloak would not be able to log the changes. However, if you want to find where any data is stored, the easiest approach is likely to create a database dump (on a test database) and then grepping for it. NB: re-reading the original message, it says that it's in the `CREDENTIALS` database.

@farnulfo, regarding the stored secret, I don't think it's possible for it to not match the data in the QR code. Maybe there is a better source of information, but the [Google Authenticator wiki](https://github.com/google/google-authenticator/wiki/Key-Uri-Format#secret) mentions that the secret value should not have any padding, so maybe that's the reason they differ? Also, the record for the OTP is probably stored after a successful TOTP code validation, so you'd need to save the QR, validate the OTP in the web form and then check the database if the secret value is the same.

**Reply by SatriyoJati** [reply]
@darius-m so, is it possible to get secret key from keycloak user? I need to create custom totp generator, so the user of keycloak that has enabled totp, has secret key. I want this secret key to be shared securely to device that use custom totp generator. in the simple word, i want to share totp secret key from user to other application, so the app can generate otp 6 digit,

**Reply by marsteel** [reply]
https://lists.jboss.org/pipermail/keycloak-dev/2017-July/009686.html

1wLkKThhSxoirVo08MCT is random string stored in DB. Convert it to bytes first and encode it with BASE32.

import base64
base64.b32encode(str.encode(“1wLkKThhSxoirVo08MCT”))
b’GF3UY22LKRUGQU3YN5UXEVTPGA4E2Q2U’

GF3UY22LKRUGQU3YN5UXEVTPGA4E2Q2U is the secret string that you can use to configure your OTP apps
